src/main/java/org/folio/service/financedata/FinanceDataService.java

```diff
@@ -45,18 +53,93 @@ public FinanceDataService(FundService fundService, BudgetService budgetService,
 
   public Future<Void> update(FyFinanceDataCollection entity, RequestContext requestContext) {
     if (CollectionUtils.isEmpty(entity.getFyFinanceData())) {
-      return Future.succeededFuture();
+      return succeededFuture();
     }
     var dbClient = requestContext.toDBClient();
+    Map<String, String> okapiHeaders = requestContext.getHeaders();
     return dbClient
-      .withTrans(conn -> {
-        var updateFundFuture = processFundUpdate(entity, conn);
-        var updateBudgetFuture = processBudgetUpdate(entity, conn);
-        return GenericCompositeFuture.all(List.of(updateFundFuture, updateBudgetFuture))
-          .compose(v -> processAllocationTransaction(entity, conn, requestContext.getHeaders()));
-      })
+      .withTrans(conn -> fiscalYearService.getFiscalYearById(getFiscalYearId(entity), conn)
+        .compose(fiscalYear -> createBudgets(entity, fiscalYear, conn, okapiHeaders)
+          .compose(v -> updateFundAndBudget(entity, conn))
+          .compose(v -> processAllocationTransaction(entity, fiscalYear, conn, okapiHeaders))))
       .onSuccess(v -> logger.info("Successfully updated finance data"))
-      .onFailure(e -> logger.error("Failed to update finance data", e))
+      .onFailure(e -> logger.error("Failed to update finance data", e));
+  }
+
+  private Future<Void> createBudgets(FyFinanceDataCollection entity, FiscalYear fiscalYear, DBConn conn,
+      Map<String, String> okapiHeaders) {
+    if (fiscalYearIsPast(fiscalYear)) {
+      return succeededFuture();
+    }
+    List<FyFinanceData> dataList = entity.getFyFinanceData().stream()
```

Comment:
as option, we can rename to `var budgetRequiredFinanceDataList` to be easier to understand

Reply:
I think `List<FyFinanceData> dataList` is easier to understand. The type is not obvious after the stream operation, so giving it explicitely helps to understand the code. A longer variable name like `budgetRequiredFinanceDataList` would also make the code harder to read, and it would not help understanding what is in that data list.